core/trino-main/src/main/java/io/trino/execution/QueryState.java

```
@@ -73,4 +73,22 @@ public boolean isDone()
     {
         return doneState;
     }
+
+    public boolean isPlanningFinished()
+    {
+        switch (this) {
+            case QUEUED:
+            case WAITING_FOR_RESOURCES:
+            case DISPATCHING:
+            case PLANNING:
+                return false;
+            case STARTING:
+            case RUNNING:
+            case FINISHING:
+            case FINISHED:
+            case FAILED:
+                return true;
+        }
+        return true;
```

this should throw an exception